skyportal/handlers/thumbnail.py

```diff
@@ -0,0 +1,136 @@
+import os
+import io
+import base64
+from pathlib import Path
+from marshmallow.exceptions import ValidationError
+from PIL import Image
+from baselayer.app.access import permissions, auth_or_token
+from .base import BaseHandler
+from ..models import DBSession, Photometry, Comment, Instrument, Source, Thumbnail
+
+
+class ThumbnailHandler(BaseHandler):
+    @permissions(['Upload data'])
+    def post(self):
+        """
+        ---
+        description: Upload thumbnails
+        parameters:
+          - in: path
+            name: source_id
+            schema:
+              type: string
+            description: ID of source associated with thumbnails. If specified, without `photometry_id`, the first photometry point associated with specified source will be associated with thumbnail(s).
+          - in: path
+            name: photometry_id
+            schema:
+              type: integer
+            description: ID of photometry to be associated with thumbnails. If omitted, `source_id` must be specified, in which case the first photometry entry associated with source will be used.
+          - in: path
+            name: data
+            schema:
+              type: string
+              format: byte
+            description: base64-encoded PNG image file contents. Image size must be between 100px and 500px on a side.
+          - in: path
+            name: ttype
+            schema:
+              type: string
+            description: Thumbnail type. Must be one of 'new', 'ref', 'sub', 'sdss', 'dr8', 'new_gz', 'ref_gz', 'sub_gz'
+        responses:
+          200:
+            content:
+              application/json:
+                schema:
+                  allOf:
+                    - Success
+                    - type: object
+                      properties:
+                        id:
+                          type: int
+                          description: New thumbnail ID
+        """
+        data = self.get_json()
+        if 'photometry_id' in data:
+            phot = Photometry.query.get(int(data['photometry_id']))
+            source = phot.source
+        elif 'source_id' in data:
+            source = Source.query.get(data['source_id'])
+            try:
+                phot = source.photometry[0]
+            except IndexError:
+                return self.error('Specified source does not yet have any photometry data.')
+        else:
+            return self.error('One of either source_id or photometry_id are required.')
+        try:
+            t = create_thumbnail(data['data'], data['ttype'], source.id, phot)
+        except Exception as e:
+            return self.error(str(e))
```

Comment:
Heh, now that I think about it: what happens when exceptions get raised from handlers?  Perhaps we already explicitly call `error` on them.  Either way, this is fine, but would be good to know!